src/liblsquic/lsquic_enc_sess_ietf.c

```diff
@@ -2450,10 +2454,10 @@ iquic_esf_decrypt_packet (enc_session_t *enc_session_p,
         enc_sess->esi_key_phase = key_phase;
     }
 
-    packet_in->pi_data_sz = packet_in->pi_header_sz + out_sz;
```

Comment:
the unencrypted buffer data len here is less than the origin encrypted data buffer len, maybe cause the old encrypted data buffer put in wrong buffer pool
packet_in->pi_data_sz(new) == packet_in->pi_header_sz + out_sz == dst_sz == packet_in->pi_data_sz(old) - 16

> I think here would never leak memory, because packet_in->pi_header_sz + out_sz == dst_sz.

Reply:
> the unencrypted buffer data len here is less than the origin encrypted data buffer len, maybe cause the old encrypted data buffer put in wrong buffer pool packet_in->pi_data_sz(new) == packet_in->pi_header_sz + out_sz == dst_sz == packet_in->pi_data_sz(old) - 16
> 
> > I think here would never leak memory, because packet_in->pi_header_sz + out_sz == dst_sz.

Maybe you can run some pressure test, and then check the engine statistics output: 
<img width="813" alt="image" src="https://user-images.githubusercontent.com/12257285/201807251-1d026519-965d-4ce7-b0d4-f1fc12f40edd.png">

Reply:
Actually, it's hard to see memory leak even in pressure test here because the packet_in data buffer is not own data in most cases.

Reply:
> Actually, it's hard to see memory leak even in pressure test here because the packet_in data buffer is not own data in most cases.

If memory leak happen, after all connections destroy, here objs_out will be a big number (unsigned integer underflow), like:
<img width="1247" alt="image" src="https://user-images.githubusercontent.com/12257285/202065273-1d5d992c-41d1-449a-8433-9ef4f092fec1.png">

Reply:
> Actually, it's hard to see memory leak even in pressure test here because the packet_in data buffer is not own data in most cases.

1) const size_t dst_sz = packet_in->pi_data_sz - 16;
2) dst = lsquic_mm_get_packet_in_buf(&enpub->enp_mm, dst_sz);
3) EVP_AEAD_CTX_open(&crypto_ctx->yk_aead_ctx,
                dst + packet_in->pi_header_sz, &out_sz,
                dst_sz - packet_in->pi_header_sz, nonce, crypto_ctx->yk_iv_sz,
                packet_in->pi_data + packet_in->pi_header_sz,
                packet_in->pi_data_sz - packet_in->pi_header_sz,
                dst, packet_in->pi_header_sz)
    Here: out_sz == dst_sz - packet_in->pi_header_sz

4) packet_in->pi_data_sz = packet_in->pi_header_sz + out_sz;
    Here: packet_in->pi_data_sz == dst_sz
5) lsquic_mm_put_packet_in_buf(&enpub->enp_mm, packet_in->pi_data,
                                                        packet_in->pi_data_sz);
    Here, packet_in->pi_data_sz same with 1) dst_sz, so here MUST not happen memory leak.

Reply:
> 
2）get dst buffer, but 5) put in packet_in->pi_data, not the dst buffer. packet_in->pi_data is allocated before this funcution

Reply:
You mean packet_in malloc in Func: lsquic_conn_copy_and_release_pi_data?

Reply:
yes, and it seems Q050 also use function lsquic_conn_copy_and_release_pi_data when the decrypt keys are not ready
mini_conn_ci_Q050_packet_in -> conn_decrypt_packet_or -> lsquic_conn_copy_and_release_pi_data

does Q050 also has a limited packet_in size 1370?

even if it does not cause memory leaks because if the limit of packet size, I think is better to assign packet_in->pi_data_sz after lsquic_mm_put_packet_in_buf

> You mean packet_in malloc in Func: lsquic_conn_copy_and_release_pi_data?
> 
> lsquic_conn_copy_and_release_pi_data only used by gquic(Q043 and Q046), and packet_in size are limited with 1370, all packet_in memory will be malloc from mm->packet_in_bufs[0] and free to mm->packet_in_bufs[0].
> 
> Could you find any other lsquic_mm_get_packet_in_buf calls with packet_in->pi_data_sz?

Reply:
> yes, and it seems Q050 also use function lsquic_conn_copy_and_release_pi_data when the decrypt keys are not ready mini_conn_ci_Q050_packet_in -> conn_decrypt_packet_or -> lsquic_conn_copy_and_release_pi_data
> 
> does Q050 also has a limited packet_in size 1370?
> 
> even if it does not cause memory leaks because if the limit of packet size, I think is better to assign packet_in->pi_data_sz after lsquic_mm_put_packet_in_buf
> 
> > You mean packet_in malloc in Func: lsquic_conn_copy_and_release_pi_data?
> > lsquic_conn_copy_and_release_pi_data only used by gquic(Q043 and Q046), and packet_in size are limited with 1370, all packet_in memory will be malloc from mm->packet_in_bufs[0] and free to mm->packet_in_bufs[0].
> > Could you find any other lsquic_mm_get_packet_in_buf calls with packet_in->pi_data_sz?

I think you are right about Q050!